### PR TITLE
[mempool] support flexible txn broadcast scheduling

### PR DESCRIPTION
**Pull request by christinacelee**
## Motivation

Originally, mempool txn broadcasts followed a rigid schedule - a fixed interval (50 ms) to all peers. There was no flexible scheduling in terms of time or recipient.

With this PR, txn broadcasts can now be scheduled at any time, and each peer can have its own broadcast schedule. 

The current implementation still follows the original protocol of fixed-interval broadcasting, but this framework lays the foundation for later supporting features that require flexible broadcast scheduling such as backpressure and responsive broadcasting

## Test Plan

Refactored and existing tests pass

### CODE REVIEW

File: mempool/src/shared_mempool/peer_manager.rs [hunk 1]
```diff
@@ -1,121 +1,148 @@
...
         }
     }
 
-    pub fn is_upstream_peer(&self, network_id: PeerId, peer_id: PeerId) -> bool {
-        self.upstream_config.is_upstream_peer(network_id, peer_id)
+    // checks whether a peer is a chosen broadcast recipient:
+    // - all primary peers
+    // - select fallback peers, if k-policy is enabled
+    // this does NOT check for whether this peer is alive
+    fn is_picked_peer(&self, peer: PeerNetworkId) -> bool {
```
Comment (phoenix-antigravity) [inline mempool/src/shared_mempool/peer_manager.rs:145]: qq: why won't try to always schedule broadcast for new_peer event (and actually even ignore lost peer event) and then check if peer is picked on a fly? 
so if it's not it's just noop routine
also you won't need to persist `picked_fallback_peers` in this case, but I'm not sure I fully understand its use

File: mempool/src/shared_mempool/peer_manager.rs [hunk 2]
```diff
@@ -1,121 +1,94 @@
...
-                .iter()
-                .filter(|(peer, state)| !picked_peers.contains_key(&peer) && state.is_alive)
-                .map(|(peer, state)| (*peer, state.clone()))
-                .choose_multiple(
-                    &mut rand::thread_rng(),
-                    self.min_broadcast_recipient_count - picked_peers_count,
-                )
-                .into_iter()
-                .collect();
+    pub fn update_peer_broadcast(&mut self, update: (PeerNetworkId, u64)) {
```
Comment (phoenix-antigravity) [inline mempool/src/shared_mempool/peer_manager.rs:52]: nit: why not 2 named parameters instead?

File: mempool/src/shared_mempool/coordinator.rs [hunk 3]
```diff
@@ -143,7 +186,12 @@ pub(crate) async fn request_coordinator<V>(
                                             .with_label_values(&["received".to_string().deref(), peer_id.to_string().deref()])
                                             .inc_by(transactions.len() as i64);
                                         let smp_clone = smp.clone();
-                                        let timeline_state = match peer_manager.is_upstream_peer(network_id, peer_id) {
+                                        let peer = PeerNetworkId(network_id, peer_id);
+                                        let timeline_state = match peer_manager
+                                            .lock()
```
Comment (phoenix-antigravity) [inline mempool/src/shared_mempool/coordinator.rs:191]: nit: would love to get rid of mutex around peer_manager,  but can we at least move it back to one level down as in previous version
In such case at least `is_upstream_peer` calls won't require a lock acquisition

File: mempool/src/shared_mempool/peer_manager.rs [hunk 4]
```diff
@@ -1,121 +1,94 @@
...
+        }
+
+        let num_alive_primary_peers = self
+            .peer_info
+            .iter()
+            .filter(|(peer, state)| self.is_primary_upstream_peer(**peer) && state.is_alive)
+            .count();
+
+        // for fallback peers, broadcast if k-policy is on
+        num_alive_primary_peers < self.min_broadcast_recipient_count
```
Comment (phoenix-antigravity) [inline mempool/src/shared_mempool/peer_manager.rs:92]: ok, so with this change we always broadcast to *all* fallback peers if num alive primaries < min_broadcast_recipient_count

btw, we probably shouldn't do that if all primaries are alive but their number is still less than min_broadcast_recipient_count. Imagine wallet use case where number of alive validators is 1, but  k for fallback is 3. You don't want to start broadcasting to secondaries right away

File: mempool/src/shared_mempool/coordinator.rs [hunk 5]
```diff
@@ -20,60 +23,90 @@ use channel::libra_channel;
...
-    let subscribers = smp.subscribers;
+    let mut scheduled_broadcasts = FuturesUnordered::new();
+    let interval_ms = smp.config.shared_mempool_tick_interval_ms;
+
+    loop {
+        tick(&mut broadcast_ticker).await;
+
+        ::futures::select! {
+            new_peer = schedule_broadcast_rx.select_next_some() => {
+                if tasks::broadcast_single_peer(new_peer, &peer_manager, &mempool, &mut network_senders, batch_size) {
```
Comment (phoenix-antigravity) [inline mempool/src/shared_mempool/coordinator.rs:65]: why not always schedule next broadcast and check on a fly if current one should be noop?
with current code, assuming primary is flaky, how fallback peer can be reliably scheduled ?

File: mempool/src/shared_mempool/coordinator.rs [hunk 6]
```diff
@@ -20,60 +23,88 @@ use channel::libra_channel;
...
-        tasks::sync_with_peers(
-            peer_manager.clone(),
-            &mempool,
-            network_senders.clone(),
-            batch_size,
-        )
-        .await;
-        notify_subscribers(SharedMempoolNotification::Sync, &subscribers);
+        ::futures::select! {
+            new_peer = schedule_broadcast_rx.select_next_some() => {
```
Comment (phoenix-antigravity) [inline mempool/src/shared_mempool/coordinator.rs:64]: nit: there's exactly same logic for handling both of channels. Can we just merge then via `select` call 

and then no need to have separate schedule_next_broadcast function; and pass local variable scheduled_broadcasts around

File: mempool/src/shared_mempool/peer_manager.rs [hunk 7]
```diff
@@ -28,94 +24,93 @@ pub(crate) struct PeerManager {
...
-                )
-                .into_iter()
-                .collect();
+    fn is_primary_upstream_peer(&self, peer: PeerNetworkId) -> bool {
+        self.upstream_config.is_primary_upstream_peer(peer)
+    }
 
-            picked_peers.extend(fallback_peers);
+    // returns a peer's sync state if the peer is a chosen broadcast peer
+    pub fn get_picked_state(&self, peer: PeerNetworkId) -> Option<PeerSyncState> {
```
Comment (phoenix-antigravity) [inline mempool/src/shared_mempool/peer_manager.rs:83]: nit: this function has interesting signature. Can we just check `is_picked_peer` on call site level and here always return PeerSyncState?

File: mempool/src/shared_mempool/coordinator.rs (hunk 6 above)
Comment (christinacelee) [reply mempool/src/shared_mempool/coordinator.rs:64]: I don't think we can fuse two streams of different types together into a single future (e.g. `FuturesUnordered` can only be fused with other `FuturesUnordered`) - `select!` is used for exactly this: combining futures of distinct types together

**Pull request closed without merging**